src/Query.php

```diff
@@ -0,0 +1,257 @@
+<?php
+/**
+ * Part of the Joomla GSoC Webservices Project
+ *
+ * @copyright  Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE
+ */
+
+namespace Joomla\Entity;
+
+use Joomla\Database\QueryInterface;
+use Joomla\Database\DatabaseDriver;
+
+/**
+ * Class Query
+ */
+class Query
+{
+	/**
+	 * The Joomla QueryInterface.
+	 *
+	 * @var QueryInterface
+	 */
+	protected $query;
+
+	/**
+	 * The connection name for the model.
+	 *
+	 * @var DatabaseDriver
+	 */
+	protected $db;
+
+	/**
+	 * The connection name for the model.
+	 *
+	 * @var Model
+	 */
+	protected $model;
+
+	/**
+	 * The methods that should be returned from Joomla Database query builder.
+	 *
+	 * @var array
+	 */
+	protected $passthru = [
+		'select', 'where', 'from'
+	];
+
+	/**
+	 * Create a new Query instance.
+	 *
+	 * @param  QueryInterface  $query
+	 * @param  DatabaseDriver $db
+	 * @param  Model $model
+	 * @return void
+	 */
+	public function __construct($query, $db, &$model)
+	{
+		$this->query = $query;
+		$this->db    = $db;
+		$this->model = $model;
+
+	}
+
+
+	/**
+	 * Inserts a single instance of a model.
+	 *
+	 * @param Model $model
+	 *
+	 * @return bool
+	 */
+	public function insert()
+	{
+		$fields       = [];
+		$values       = [];
+
+		// Iterate over the object variables to build the query fields and values.
+		foreach ($this->model->getAttributes() as $k => $v)
+		{
+			// Prepare and sanitize the fields and values for the database query.
+			$fields[] = $this->db->quoteName($k);
+			$values[] = $this->db->quote($v);
+		}
+
+		// Create the base insert statement.
+		$this->query->insert($this->db->quoteName($this->model->getTable()))
+			->columns($fields)
+			->values(implode(',', $values));
+
+		// Set the query and execute the insert.
+		$success = $this->db->setQuery($this->query)->execute();
+
+		if ($this->model->isIncrementing())
+		{
+			$id = $this->db->insertid();
+			$key = $this->model->getPrimaryKey();
+			// Update the primary key if it exists.
+			if ($key && $id && is_string($key))
+			{
+				$this->model->setPrimaryKey($id);
```

Comment:
This looks wrong to me - you're setting the primary key column to be the value of the primary key column...